src/gcrypt/rsagen.c

```diff
@@ -20,220 +20,299 @@
 #include "system.h"
 
 #include <gcrypt.h>
+#include <assert.h>
+
+#include "../rsagen.h"
+#include "xalloc.h"
+#include "rsa.h"
+#include "pem.h"
+
+// ASN.1 tags.
+typedef enum {
+	TAG_INTEGER = 2,
+	TAG_SEQUENCE = 16,
+} asn1_tag_t;
+
+static size_t der_tag_len(size_t n) {
```

Comment:
This is why I really dislike gcrypt. They don't support PEM, they have those stupid s-expressions... it doesn't help security if you force users to implement their own parsers for bog standard crypto file formats :(

Reply:
Oh I just saw that there are some companion libraries to libgcrypt, one of them is [libksba](https://gnupg.org/software/libksba/index.html), which might be able to do the PEM decoding.

Reply:
I don't think it helps with PEM encoding or decoding.

Its [reader](https://github.com/gpg/libksba/blob/88392670f4e5640744b8bd6a68debd2e05274d19/src/reader.c) and [writer](https://github.com/gpg/libksba/blob/88392670f4e5640744b8bd6a68debd2e05274d19/src/writer.c) are simple wrappers around `fread()` that don't do much (you still have to provide base64 encoding/decoding, write out PEM headers, etc. That's what GPG does: [read](https://github.com/gpg/gnupg/blob/283ccbc824d8062b86e27818a84ab5a29facd7a5/common/ksba-io-support.c#L177), [write](https://github.com/gpg/gnupg/blob/283ccbc824d8062b86e27818a84ab5a29facd7a5/common/ksba-io-support.c#L427).)

It seems to have a pretty decent DER builder. I almost finished porting the code and only then noticed that it's [less than 1.5 years](https://github.com/gpg/libksba/commit/cf49d3e60a67180fcb1b9005d910f015b388cf3c) old and is available on relatively few platforms right now:

- Debian stable
- Fedora (latest)
- Ubuntu (latest)
- FreeBSD 13
- OpenBSD 6.9
- NetBSD 9.2

but not:

- Debian oldstable
- Ubuntu 20.04 (current LTS)
- RHEL 8
- OpenBSD 6.8
- NetBSD 8.2

Any thoughs? Should I keep the hand-written DER encoder and use ksba through conditional compilation, or do we declare libgcrypt as supported only on recent systems?

Reply:
Hm, let's keep the hand-written code then, and move to the external library when it's more mature.